File: src/net/eduard/api/lib/abstraction/Minecraft_v1_7_R4.java
package net.eduard.api.lib.abstraction;

import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_7_R4.entity.CraftPlayer;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import net.minecraft.server.v1_7_R4.EntityPlayer;
import net.minecraft.server.v1_7_R4.EnumClientCommand;
import net.minecraft.server.v1_7_R4.MathHelper;
import net.minecraft.server.v1_7_R4.Packet;
import net.minecraft.server.v1_7_R4.PacketPlayInClientCommand;
import net.minecraft.server.v1_7_R4.PacketPlayOutEntityDestroy;
import net.minecraft.server.v1_7_R4.PacketPlayOutEntityHeadRotation;
import net.minecraft.server.v1_7_R4.PacketPlayOutEntityMetadata;
import net.minecraft.server.v1_7_R4.PacketPlayOutNamedEntitySpawn;
import net.minecraft.server.v1_7_R4.PacketPlayOutPlayerInfo;

/**
 * @author Eduard
 */
public class Minecraft_v1_7_R4 extends Minecraft {

    public void sendPacket(Object packet, Player player) {
        ((CraftPlayer) player).getHandle().playerConnection.sendPacket((Packet) packet);
    }

    @Override
    public void removeFromTab(Player playerRemoved) {

        final PacketPlayOutPlayerInfo removePlayerInfo = PacketPlayOutPlayerInfo
                .removePlayer(((CraftPlayer) playerRemoved).getHandle());

        sendPacketsToOthers(playerRemoved, removePlayerInfo);
    }

    @Override
    public void addToTab(Player playerToAdd) {
        final PacketPlayOutPlayerInfo addPlayerInfo = PacketPlayOutPlayerInfo
                .addPlayer(((CraftPlayer) playerToAdd).getHandle());
        final PacketPlayOutPlayerInfo updatePlayerInfo = PacketPlayOutPlayerInfo
                .updateDisplayName(((CraftPlayer) playerToAdd).getHandle());
        sendPacketsToOthers(playerToAdd, addPlayerInfo, updatePlayerInfo);

    }

    @Override
    public Object getItemNBT(ItemStack item) {

        return null;
    }

    @Override
    public ItemStack setItemNBT(ItemStack item, Object nbt) {

        return item;
    }

    @Override
    public void respawnPlayer(Player playerToRespawn) {
        final EntityPlayer entityPlayer = ((CraftPlayer) playerToRespawn).getHandle();
        final PacketPlayOutEntityDestroy destroy = new PacketPlayOutEntityDestroy(new int[]{entityPlayer.getId()});
        final PacketPlayOutNamedEntitySpawn spawn = new PacketPlayOutNamedEntitySpawn(entityPlayer);
        final PacketPlayOutPlayerInfo addPlayerInfo = PacketPlayOutPlayerInfo
                .addPlayer(((CraftPlayer) playerToRespawn).getHandle());
        final PacketPlayOutEntityMetadata metadata = new PacketPlayOutEntityMetadata(entityPlayer.getId(),
                entityPlayer.getDataWatcher(), true);
        final PacketPlayOutEntityHeadRotation headRotation = new PacketPlayOutEntityHeadRotation(entityPlayer,
                (byte) MathHelper.d(entityPlayer.getHeadRotation() * 256.0f / 360.0f));
        final PacketPlayOutPlayerInfo removePlayerInfo = PacketPlayOutPlayerInfo
                .removePlayer(((CraftPlayer) playerToRespawn).getHandle());

        sendPacketsToOthers(playerToRespawn, removePlayerInfo, addPlayerInfo, destroy, spawn, metadata, headRotation);

    }

    @Override
    public void performRespawn(Player player) {
        ((CraftPlayer) player).getHandle().playerConnection
                .a(new PacketPlayInClientCommand(EnumClientCommand.PERFORM_RESPAWN));

    }

    @Override
    public void setHeadSkin(ItemStack head, String texture, String signature) {

    }

    @Override
    public void setPlayerName(Player player, String newName) {

    }

    @Override
    public void setPlayerSkin(Player player, String newSkin) {


    }

    @Override
    public void sendActionBar(Player player, String message) {


    }

    @Override
    public void sendParticle(Player player, String name, Location location, int amount, float xOffset, float yOffset, float zOffset, float speed) {

    }

}

File: MineNMS-1_7/src/main/java/Minecraft_v1_7_R4.kt
package net.eduard.api.lib.abstraction

import net.minecraft.server.v1_7_R4.*
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.Chest
import org.bukkit.craftbukkit.v1_7_R4.entity.CraftPlayer
import org.bukkit.entity.Creature
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

/**
 * @author Eduard
 */
class Minecraft_v1_7_R4 : Minecraft() {

    override fun setBlock(block: Block, chunk: Chunk, material: Material, data: Int, updateLightning: Boolean) {

    }
    override fun sendPacket(packet: Any, player: Player) {
        (player as CraftPlayer).handle.playerConnection
            .sendPacket(packet as Packet)
    }

    override fun forceOpen(chest: Chest, player: Player) {

    }

    override fun canAttackMelee(creature: Creature, classEntityName: String, priority: Int) {}
    override fun followTarget(creature: Creature) {
    }

    override fun followLocation(creature: Creature, location: Location) {
    }

    override fun canTarget(creature: Creature, classEntityName: String, priority: Int) {}

    override fun removeGoals(creature: Creature) {}

    override fun removeTargetGoals(creature: Creature) {}
    override fun removeFromTab(playerRemoved: Player) {
        val removePlayerInfo = PacketPlayOutPlayerInfo
            .removePlayer((playerRemoved as CraftPlayer).handle)
        sendPacketsToOthers(playerRemoved, removePlayerInfo)
    }

    override fun addToTab(playerToAdd: Player) {
        val addPlayerInfo = PacketPlayOutPlayerInfo
            .addPlayer((playerToAdd as CraftPlayer).handle)
        val updatePlayerInfo = PacketPlayOutPlayerInfo
            .updateDisplayName(playerToAdd.handle)
        sendPacketsToOthers(playerToAdd, addPlayerInfo, updatePlayerInfo)
    }

    override fun updateDisplayName(playerToAdd: Player) {}
    override fun getItemNBT(item: ItemStack): Any {
        return 1
    }

    override fun setItemNBT(item: ItemStack, nbt: Any): ItemStack {
        return item
    }

    override fun disableAI(entity: Entity) {}
    override fun respawnPlayer(playerToRespawn: Player) {
        val entityPlayer = (playerToRespawn as CraftPlayer).handle
        val destroy = PacketPlayOutEntityDestroy(*intArrayOf(entityPlayer.id))
        val spawn = PacketPlayOutNamedEntitySpawn(entityPlayer)
        val addPlayerInfo = PacketPlayOutPlayerInfo
            .addPlayer(playerToRespawn.handle)
        val metadata = PacketPlayOutEntityMetadata(
            entityPlayer.id,
            entityPlayer.dataWatcher, true
        )
        val headRotation = PacketPlayOutEntityHeadRotation(
            entityPlayer,
            MathHelper.d(entityPlayer.headRotation * 256.0f / 360.0f).toByte()
        )
        val removePlayerInfo = PacketPlayOutPlayerInfo
            .removePlayer(playerToRespawn.handle)
        sendPacketsToOthers(playerToRespawn, removePlayerInfo, addPlayerInfo, destroy, spawn, metadata, headRotation)
    }

    override fun reloadPlayer(player: Player) {}
    override fun setTabList(player: Player, header: String, footer: String) {
        TODO("Not yet implemented")
    }

    override fun performRespawn(player: Player) {
        (player as CraftPlayer).handle.playerConnection
            .a(PacketPlayInClientCommand(EnumClientCommand.PERFORM_RESPAWN))
    }

    override fun setHeadSkin(head: ItemStack, texture: String, signature: String) {}
    override fun setPlayerName(player: Player, newName: String) {}
    override fun setPlayerSkin(player: Player, newSkin: String) {}
    override fun sendActionBar(player: Player, message: String) {}
    override fun sendParticle(
        player: Player,
        name: String,
        location: Location,
        amount: Int,
        xOffset: Float,
        yOffset: Float,
        zOffset: Float,
        speed: Float
    ) {
    }

    override fun getPlayers(): Collection<Player> {
        return mutableListOf()
    }
}